rest_api/sawtooth_rest_api/routes.py

```diff
@@ -178,56 +139,60 @@ def _try_validator_request(self, message_type, content):
         try:
             response = future.result(timeout=timeout)
         except FutureTimeoutError:
-            raise web.HTTPServiceUnavailable(reason=timeout_msg)
+            raise errors.ValidatorUnavailable()
 
         return response.content
 
     @staticmethod
-    def _try_response_parse(proto, response):
+    def _try_response_parse(proto, response, traps=None):
         """
         Parses a protobuf response from the validator
         Raises common validator error statuses as HTTP errors
         """
-        unknown_msg = 'An unknown error occured with your request'
-        notfound_msg = 'There is no resource at that root, address or prefix'
-
         parsed = proto()
         parsed.ParseFromString(response)
+        traps = traps or []
 
+        # Not every protobuf has every status enum, so pass AttributeErrors
+        try:
+            traps.append(errors.Unknown(proto.ERROR))
+        except AttributeError:
+            pass
         try:
-            if parsed.status == proto.ERROR:
-                raise web.HTTPInternalServerError(reason=unknown_msg)
-            if parsed.status == proto.NORESOURCE:
-                raise web.HTTPNotFound(reason=notfound_msg)
+            traps.append(errors.MissingGenesis(proto.NOGENESIS))
         except AttributeError:
-            # Not every protobuf has every status, so pass AttributeErrors
             pass
 
+        for trap in traps:
+            trap.check(parsed.status)
+
         return MessageToDict(parsed, preserving_proto_field_name=True)
 
     @staticmethod
-    def _wrap_response(data=None, head=None, link=None):
+    def _safe_get(obj, key, default=''):
+        """
+        aiohttp very helpfully parses param strings to replace '+' with ' '
+        This is very bad when your block ids contain meaningful +'s
+        """
+        return obj.get(key, default).replace(' ', '+')
```

Comment:
Ah. I'm not sure I'm a fan of that. Feels like swapping one hacky work around for another. We'd be lying to the client about what the block ids are, so the block ids we give them would _only_ be usable through the Rest Api. They wouldn't match the protobuf binaries that got posted through `POST /batches` either.

Also, the query string is the least of the issue. That you _can_ use percent escaping for, for whatever reason. Also, it would be easy enough to parse those manually. But `match_info` is giving us path parameters that that we named when we were setting up the routes. The only way to manually parse those would either be _very_ hacky, or very low level, some sort of middleware I don't entirely know how to build at the moment.